README.md
# simple-webserver

[![Build Status](https://travis-ci.org/andygrunwald/simple-webserver.svg?branch=master)](https://travis-ci.org/andygrunwald/simple-webserver)
[![GoDoc](https://godoc.org/github.com/andygrunwald/simple-webserver?status.svg)](https://godoc.org/github.com/andygrunwald/simple-webserver)
[![Go Report Card](https://goreportcard.com/badge/github.com/andygrunwald/simple-webserver)](https://goreportcard.com/report/github.com/andygrunwald/simple-webserver)

A small webserver (written in [Go](http://golang.org/)) for testing various technologies, techniques and concepts like [Docker](https://www.docker.com/), [Marathon](https://mesosphere.github.io/marathon/) / [Apache Mesos](http://mesos.apache.org/), [Kubernetes](http://kubernetes.io/), [API Blueprint](https://apiblueprint.org/) and others.

All guides how to use, play and use a technique or concept are documented here. Feel free to play around and learn as much as i do.

---

## Table of Contents

1. [Batteries Included](#batteries-included)
2. [Why this project?](#why-this-project)
3. [Build instructions](#build-instructions)
	1. [Native application](#native-application)
	2. [Docker](#docker)
4. [Tryout everything](#tryout-everything)
	1. [Native application](#native-application-1)
	2. [Docker](#docker-1)
	3. [Docker Compose](#docker-compose)
	3. [Marathon (native application)](#marathon-native-application)
	4. [Marathon (docker container)](#marathon-docker-container)
	5. [Marathon incl. Redis backend (docker container)](#marathon-incl-redis-backend-docker-container)
	6. [Kubernetes](#kubernetes)
	7. [API Blueprint: Validating API description](#api-blueprint-validating-api-description)
	8. [API Blueprint: Generating API docs](#api-blueprint-generating-api-docs)
5. [Things on the list to try](#things-on-the-list-to-try)
6. [Contact](#contact)
7. [License](#license)

## Batteries Included

Here you will find a list of tested technologies, techniques and concepts within this project:

* [Go](http://golang.org/): How to write a small webserver with a minimal [Redis](http://redis.io/) backend in this programming language (incl. [GoDoc](https://godoc.org/github.com/andygrunwald/simple-webserver) and [unit tests](./main_test.go))
* [12factor](https://12factor.net/config): Configure an application via environment variables
* [Travis CI](https://travis-ci.org/): Execute unit tests on every push and pull request
* [Docker](https://www.docker.com/): How to isolate a Go app in a docker container
* [Docker Compose](https://docs.docker.com/compose/): How to start a multi-container setup with one command
* [Marathon](https://mesosphere.github.io/marathon/) @ [Apache Mesos](http://mesos.apache.org/): How to deploy this app on a Marathon cluster (native and in a docker container)
* [Kubernetes](http://kubernetes.io/): How to deploy this app on a Kubernetes cluster
* [API Blueprint](https://apiblueprint.org/): API development based on markdown specifications inclusive API doc generation by [aglio](https://github.com/danielgtaylor/aglio) and validation by [dredd](https://github.com/apiaryio/dredd)

## Why this project?

The information technology and computer science world is getting crazy.
Every day new tooling will be released, new techniques comes up and some concepts are getting best practice.

**It is important to stay up to date and this is the really hard part here.**

This project is my personal playground to try out all the new things i am excited of.
I am aware that this project is not representative and comparable with a real products that solves problems of the world.
Those projects are much more complex and have several other dependencies.
But the simplicity of this projects is on purpose and enables me to get a first idea and feeling of the topic i want to try in a small amount of time.

## Build instructions

### Native application

For building the native webserver you need a running [Go installation](https://golang.org/doc/install).

```sh
$ go build -o simple-webserver .
```

To compile it for a different OS (e.g. linux) use

```sh
$ GOOS=linux go build -o simple-webserver .
```

Precompiled binaries for various operating systems can be found in our [releases](https://github.com/andygrunwald/simple-webserver/releases).

If you want to execute the unit tests, run

```sh
$ go test -v ./...
```

### Docker

For building the docker image you need a running [Docker engine](https://docs.docker.com/engine/installation/).

```sh
$ docker build -t andygrunwald/simple-webserver .
```

Or you use the existing images from [Docker Hub](https://hub.docker.com/r/andygrunwald/simple-webserver/).

## Tryout everything

### Native application

The general usage of *simple-webserver* is

```sh
$ ./simple-webserver -help
Usage of ./simple-webserver:
  -listen string
    	Address + Port to listen on. Format ip:port. Environment variable: SIMPLE_WEBSERVER_LISTEN (default ":8082")
  -redis string
    	Address + Port where a redis server is listening. Environment variable: SIMPLE_WEBSERVER_REDIS (default ":6379")
```

Start the webserver with

```sh
$ ./simple-webserver
2016/06/05 14:57:18 Starting webserver and listen on :8082
```

and send your first request

```sh
$ curl http://localhost:8082/version
simple webserver v1.1.1
```

If you have a Redis server available, you can apply it via command line flag:

```sh
$ ./simple-webserver -redis ":6379"
2016/06/05 14:57:18 Starting webserver and listen on :8082
```

and call a `/ping` request

```sh
$ curl http://localhost:8082/ping
PONG
```

### Docker

Start the container with

```sh
$ docker run -d -p 8082:8082 andygrunwald/simple-webserver
9b20babef443420bf1728583fbfba......
```

and send your first request

```sh
$ curl http://DOCKER-IP:8082/version
simple webserver v1.1.1
```

Or launch *simple-webserver* with a Redis backend

```sh
$ docker run --name simple-webserver-redis -d redis
$ docker run -d -p 8082:8082 \
				--link simple-webserver-redis:redis \
				andygrunwald/simple-webserver -redis "redis:6379"
```

For further information the docker image is available at [Docker Hub](https://hub.docker.com/r/andygrunwald/simple-webserver/).

### Docker Compose

Fire up *simple-webserver* with a Redis backend with

```sh
$ docker-compose up -d
```

and call a `/ping` request

```sh
$ curl http://DOCKER-IP:8082/ping
PONG
```

### Marathon (native application)

Start the native application on a [Marathon](https://github.com/mesosphere/marathon) cluster (scheduling framework for [Apache Mesos](http://mesos.apache.org/) with

```sh
$ curl -X POST -H "Content-Type: application/json" http://marathon:8080/v2/apps -d@marathon.json
{
    "id": "/simple-webserver",
    ...
}
```

Please visit your Marathon UI at http://marathon:8080/ to see if the deployment works out.

If you use Marathon in combination with [haproxy-marathon-bridge](https://open.mesosphere.com/tutorials/service-discovery/) or another [service discovery](https://mesosphere.github.io/marathon/docs/service-discovery-load-balancing.html) solution your service is available via the **service port 11000**

```sh
$ curl http://marathon-node:11000/version
simple webserver v1.1.1
```

If you don`t have a Mesos / Marathon cluster, i highly recommend [bobrik/mesos-compose](https://github.com/bobrik/mesos-compose).

### Marathon (docker container)

Start the docker container on a [Marathon](https://github.com/mesosphere/marathon) cluster (scheduling framework for [Apache Mesos](http://mesos.apache.org/) with

```sh
$ curl -X POST -H "Content-Type: application/json" http://marathon:8080/v2/apps -d@marathon-docker.json
{
    "id":"/simple-webserver-docker",
    ...
}
```

Please visit your Marathon UI at http://marathon:8080/ to see if the deployment works out.

If you use Marathon in combination with [haproxy-marathon-bridge](https://open.mesosphere.com/tutorials/service-discovery/) or another [service discovery](https://mesosphere.github.io/marathon/docs/service-discovery-load-balancing.html) solution your service is available via the **service port 11001**

```sh
$ curl http://marathon-node:11001/version
simple webserver v1.1.1
```

If you don`t have a Mesos / Marathon cluster, i highly recommend [bobrik/mesos-compose](https://github.com/bobrik/mesos-compose).

### Marathon incl. Redis backend (docker container)

Start the docker container with a Redis backend on a [Marathon](https://github.com/mesosphere/marathon) cluster (scheduling framework for [Apache Mesos](http://mesos.apache.org/) with

```sh
$ curl -X POST -H "Content-Type: application/json" http://marathon:8080/v2/groups -d@marathon-redis-docker.json
{
    "version":"2016-06-18T16:01:45.739Z",
    ...
}
```

Please visit your Marathon UI at http://marathon:8080/ to see if the deployment works out.

If it doesn't work out, please check `constraints` value in [marathon-redis-docker.json](./marathon-redis-docker.json) if this is equal the IP of your Mesos slave. If not, adjust it accordingly.

If you use Marathon in combination with [haproxy-marathon-bridge](https://open.mesosphere.com/tutorials/service-discovery/) or another [service discovery](https://mesosphere.github.io/marathon/docs/service-discovery-load-balancing.html) solution your service is available via the **service port 11001**

```sh
$ curl http://marathon-node:11001/ping
PONG
```

If you don`t have a Mesos / Marathon cluster, i highly recommend [bobrik/mesos-compose](https://github.com/bobrik/mesos-compose).

### Kubernetes

In this section we assume you have already a running Kubernetes cluster incl. a configured [kubectl](http://kubernetes.io/docs/user-guide/prereqs/). Otherwise i recommend to read [Getting Started With A Local Kubernetes Environment](https://blog.giantswarm.io/getting-started-with-a-local-kubernetes-environment/) or [Installing/Setting Up Kubernetes @ Kubernetes Docs](http://kubernetes.io/docs/).

In the [kubernetes.yaml](./kubernetes.yaml) you will find definitions for [Deployments](http://kubernetes.io/docs/user-guide/deployments/), [Services](http://kubernetes.io/docs/user-guide/services/) and an [Ingress Resource](http://kubernetes.io/docs/user-guide/ingress/).
This definition will start the *simple-webserver* incl. a Redis instance.

So lets start.
Apply the definition:

```sh
$ kubectl apply --filename kubernetes.yaml
service "simple-webserver" created
deployment "simple-webserver" created
service "simple-webserver-redis" created
deployment "simple-webserver-redis" created
ingress "simple-webserver" created
```

Check if all deployments are there:

```sh
$ kubectl get deployment -l app=simple-webserver
NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
simple-webserver         1         1         1            1           24s
simple-webserver-redis   1         1         1            1           24s
```

Check the services:

```sh
$ kubectl get svc -l app=simple-webserver
NAME                     CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
simple-webserver         172.31.0.20    <nodes>       8082:30658/TCP   51s
simple-webserver-redis   172.31.0.238   <none>        6379/TCP         50s
```

And have a look at your [pods](http://kubernetes.io/docs/user-guide/pods/):

```
$ kubectl get pods -l app=simple-webserver
NAME                                      READY     STATUS    RESTARTS   AGE
simple-webserver-494066648-8dkhm          1/1       Running   0          1m
simple-webserver-redis-1815988752-09qog   1/1       Running   0          1m
```

Depending on your configuration *simple-webserver* should be accessible under your public ip / address via the Ingress resource.

To clean it up and delete it, fire:

```sh
$ kubectl delete service,deployment,ingress simple-webserver
service "simple-webserver" deleted
deployment "simple-webserver" deleted
ingress "simple-webserver" deleted
$ kubectl delete service,deployment simple-webserver-redis
service "simple-webserver-redis" deleted
deployment "simple-webserver-redis" deleted
```

### API Blueprint: Validating API description

[API Blueprint](https://apiblueprint.org/) is a powerful high-level API description language for web APIs.
Based on out definition file [api-blueprint.apib](./api-blueprint.apib) we are able to ...

... validate our API implementation (in *simple-webserver*) based on our API description / apib file with [dredd](https://github.com/apiaryio/dredd).
dredd will analyze the description and generate HTTP tests and fire them against a running instance of *simple-webserver*.

At first start an instance of simple-webserver. You can choose your favorite way. Native, Docker, Marathon, etc. Doesn't matter.

After this, run *dredd* tests in a docker container:

```sh
$ docker run -it --rm \
				--net host \
				-v /HOST/PATH/TO/simple-webserver:/simple-webserver \
				apiaryio/dredd \
				dredd simple-webserver/api-blueprint.apib http://INSTANCE:8082
info: Beginning Dredd testing...
pass: HEAD / duration: 30ms
...
```

Please replace

* `/HOST/PATH/TO/simple-webserver` with an absolute path to a local copy of *simple-webserver* on your machine
* `http://INSTANCE:8082` with IP:Port of a running instance of *simple-webserver*

### API Blueprint: Generating API docs

[API Blueprint](https://apiblueprint.org/) is a powerful high-level API description language for web APIs.
Based on out definition file [api-blueprint.apib](./api-blueprint.apib) we are able to ...

... generate our API documentation with [aglio](https://github.com/danielgtaylor/aglio).

```sh
$ docker run -ti --rm \
				-v `pwd`:/docs \
				humangeo/aglio -i api-blueprint.apib -o /docs/output.html
```

After this you can will find a new file `output.html` in the current directory.
Open it and you should see you API documentation.

## Things on the list to try

* [Blue-Green Deployment with Marathon](https://mesosphere.github.io/marathon/docs/blue-green-deploy.html)

## Contact

You are more than welcome to [open issues](https://github.com/andygrunwald/simple-webserver/issues) and / or [send pull requests](https://github.com/andygrunwald/simple-webserver/pulls) if you found a typo, a bug, need help to try something or want to request a new feature.

**If you appreciate this project please feel free to drop me a line and tell me!*
It's always nice to hear from people who have benefitted from this work and pushes the motivation again.**

* Twitter: [@andygrunwald](https://twitter.com/andygrunwald)
* Email: Checkout the mail at [my github profile](https://github.com/andygrunwald)

## License

This project is released under the terms of the [MIT license](http://en.wikipedia.org/wiki/MIT_License).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added

### Fixed

## [v1.1.1] - 2017-06-06

### Added

* Added log output logging to HTTP-Endpoint `/payload`

### Fixed

* Set version to v1.1.1 in all definitions

## [v1.1.0] - 2017-06-06

### Added

* HTTP-Endpoint `/payload` to debug HTTP calls
* Disk-Attribute to marathon.json (native app)
* Disk-Attribute to marathon-docker.json (docker app)
* New deployment method: Marathon incl. Redis backend (docker container)
* Table of contents
* New deployment method: Docker Compose
* New technique: API Blueprint: Validating API description
* New technique: API Blueprint: Generating API docs

### Fixed

* Changed Marathon definitions to new v1.0.0 version
* EXPOSE-Port of Dockerfile (from 8080 to 8082)
* New lines after body for `/ping` and `/version`

## [v1.0.0] - 2016-06-18

### Added

* JSON definition to deploy simple-webserver to Marathon (based on Apache Mesos)
* Dockerfile to deploy simple-webserver with Docker
* JSON definition to deploy simple-webserver in a Docker container to Marathon
* HTTP-Endpoint `/version` that will respond with the Name and Version of this app with HTTP status code 200
* HTTP-Endpoint `/kill` that will kill the webserver
* HTTP request logging to stdout
* This changelog
* "Contact" section to README
* Recommendation for bobrik/mesos-compose to README
* GoDoc
* Unit tests
* TravisCI
* "Things on the list to try" section to README
* Redis as /ping backend
* Instructions for native and Docker based start with a Redis backend

### Changed

* Deployment for Marathon Docker to Port 11001
* Minimized hardware req. for simple-webserver (native) on Marathon
* Removed \n from pong response

### Fixed

* Major refactoring (incl. more documentation) of README.md

## [v0.0.1] - 2015-09-09

### Added

* HTTP-Endpoint `/ping` that will respond static `pong` with HTTP status code 200
* HTTP-Endpoint `/` that redirects to `/ping` with HTTP status code 303 See Other
* `-listen` flag to define IP and Port for HTTP server binding

[Unreleased]: https://github.com/andygrunwald/simple-webserver/compare/v1.0.0...HEAD
[v1.0.0]: https://github.com/andygrunwald/simple-webserver/releases/tag/v1.0.0
[v0.0.1]: https://github.com/andygrunwald/simple-webserver/releases/tag/v0.0.1


main_test.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"os/exec"
	"testing"
)

type dummyStorage struct{}

func (s *dummyStorage) Ping() (string, error) {
	return "PONG", nil
}

func TestPingHandler(t *testing.T) {
	req, err := http.NewRequest("GET", "/ping", nil)
	if err != nil {
		t.Fatal(err)
	}

	backend := &dummyStorage{}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(PingHandler(backend))
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusOK {
		t.Errorf("Wrong status code: got %v, expected %v", status, http.StatusOK)
	}

	expected := "PONG\n"
	if rr.Body.String() != expected {
		t.Errorf("Unexpected body: got %q, want %q", rr.Body.String(), expected)
	}
}

type dummyErrorStorage struct{}

func (s *dummyErrorStorage) Ping() (string, error) {
	return "", fmt.Errorf("Connection timeout to storage backend")
}

func TestPingHandler_Error(t *testing.T) {
	req, err := http.NewRequest("GET", "/ping", nil)
	if err != nil {
		t.Fatal(err)
	}

	backend := &dummyErrorStorage{}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(PingHandler(backend))
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusInternalServerError {
		t.Errorf("Wrong status code: got %v, expected %v", status, http.StatusInternalServerError)
	}

	expected := "Connection timeout to storage backend"
	if rr.Body.String() != expected {
		t.Errorf("Unexpected body: got %q, want %q", rr.Body.String(), expected)
	}
}

func TestVersionHandler(t *testing.T) {
	req, err := http.NewRequest("GET", "/version", nil)
	if err != nil {
		t.Fatal(err)
	}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(VersionHandler)
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusOK {
		t.Errorf("Wrong status code: got %v, expected %v", status, http.StatusOK)
	}

	expected := fmt.Sprintf("%s v%s\n", Name, Version)
	if rr.Body.String() != expected {
		t.Errorf("Unexpected body: got %q, want %q", rr.Body.String(), expected)
	}
}

func TestKillHandler(t *testing.T) {
	// Here we only test if the end point kills itself.
	if os.Getenv("KILL_ENDPOINT") == "1" {
		req, err := http.NewRequest("DELETE", "/kill", nil)
		if err != nil {
			t.Fatal(err)
		}

		rr := httptest.NewRecorder()
		handler := http.HandlerFunc(KillHandler)
		handler.ServeHTTP(rr, req)
		return
	}

	cmd := exec.Command(os.Args[0], "-test.run=TestKillHandler")
	cmd.Env = append(os.Environ(), "KILL_ENDPOINT=1")
	err := cmd.Run()
	if e, ok := err.(*exec.ExitError); ok && !e.Success() {
		return
	}
	t.Errorf("Process ran with err %v, want exit status 1", err)
}

func TestKillHandler_WrongMethod(t *testing.T) {
	req, err := http.NewRequest("POST", "/kill", nil)
	if err != nil {
		t.Fatal(err)
	}

	rr := httptest.NewRecorder()
	handler := http.HandlerFunc(KillHandler)
	handler.ServeHTTP(rr, req)

	if status := rr.Code; status != http.StatusMethodNotAllowed {
		t.Errorf("Wrong status code: got %v, expected %v", status, http.StatusMethodNotAllowed)
	}
}

func TestEnvOrDefault_Fallback(t *testing.T) {
	expected := "dummy"
	got := EnvOrDefault("TEST_DEFAULT_ENV", expected)
	if expected != got {
		t.Errorf("Wrong fallback value: got %v, expected %v", got, expected)
	}
}

func TestEnvOrDefault_WithEnv(t *testing.T) {
	expected := "dummy-from-env"
	os.Setenv("TEST_SET_ENV", expected)
	got := EnvOrDefault("TEST_SET_ENV", "dummy-fallback")
	if expected != got {
		t.Errorf("Wrong env value: got %v, expected %v", got, expected)
	}
}


doc.go
// A small webserver for testing various technologies, techniques and concepts.
//
// This project enables you to learn new technologies by try them in
// a small and simple environment to get a first idea how things are working.
//
// This part is responsible to show the standard way of documentation
// of golang. With the idea to inline the documentation, code and docs are one.
// This minimize the risk that documentation get out of date quickly.
//
// Notice that this is only a a small and not production ready demo.
// If you want to deep dive into the topic of golang + documentation
// i suggest to checkout:
//
// https://blog.golang.org/godoc-documenting-go-code
//
// https://godoc.org/golang.org/x/tools/cmd/godoc
//
// https://github.com/fluhus/godoc-tricks
//
// https://github.com/golang/gddo
//
// If you have any suggestion or comment, please feel free to open an issue on
// this the GitHub page of this project!
//
// More information and details can be found there as well.
// Checkout https://github.com/andygrunwald/simple-webserver
//
package main


redis.go
package main

import (
	"time"

	"github.com/garyburd/redigo/redis"
)

// Storage is the basic interface to communicate
// with different storage endpoints.
type Storage interface {
	// Ping will "ping" the storage backend.
	// The function can use to check the connection from
	// the app to the storage backend.
	Ping() (string, error)
}

// RedisStorage reflects the Storage with
// a Redis NoSQL Server as backend
// See http://redis.io/
type RedisStorage struct {
	connectionPool *redis.Pool
}

// NewRedisStorage will return a new Storage type
// with Redis as backend
func NewRedisStorage(server string) Storage {
	return &RedisStorage{
		connectionPool: &redis.Pool{
			MaxIdle:     3,
			IdleTimeout: 240 * time.Second,
			Dial: func() (redis.Conn, error) {
				c, err := redis.Dial("tcp", server)
				if err != nil {
					return nil, err
				}
				return c, err
			},
			TestOnBorrow: func(c redis.Conn, t time.Time) error {
				_, err := c.Do("PING")
				return err
			},
		},
	}
}

// Ping will "ping" the storage backend.
// The function can use to check the connection from
// the app to the storage backend.
func (r *RedisStorage) Ping() (string, error) {
	conn := r.connectionPool.Get()
	defer conn.Close()

	res, err := redis.String(conn.Do("PING"))
	return res, err
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"io/ioutil"

	"github.com/gorilla/handlers"
)

const (
	// Name of the application
	Name = "simple webserver"
	// Version of the application
	Version = "1.1.1"
)

func main() {
	var (
		listenFlag = flag.String("listen", EnvOrDefault("SIMPLE_WEBSERVER_LISTEN", ":8082"), "Address + Port to listen on. Format ip:port. Environment variable: SIMPLE_WEBSERVER_LISTEN")
		redisFlag  = flag.String("redis", EnvOrDefault("SIMPLE_WEBSERVER_REDIS", ":6379"), "Address + Port where a redis server is listening. Environment variable: SIMPLE_WEBSERVER_REDIS")
	)
	flag.Parse()

	// Create Redis storage
	r := NewRedisStorage(*redisFlag)

	// Define HTTP endpoints
	s := http.NewServeMux()
	s.HandleFunc("/", RootHandler)
	s.HandleFunc("/ping", PingHandler(r))
	s.HandleFunc("/kill", KillHandler)
	s.HandleFunc("/version", VersionHandler)
	s.HandleFunc("/payload", PayloadHandler)

	// Bootstrap logger
	logger := log.New(os.Stdout, "", log.LstdFlags)
	logger.Printf("Starting webserver and listen on %s", *listenFlag)

	// Start HTTP Server with request logging
	loggingHandler := handlers.LoggingHandler(os.Stdout, s)
	log.Fatal(http.ListenAndServe(*listenFlag, loggingHandler))
}

// RootHandler handles requests to the "/" path.
// It will redirect the request to /ping with a 303 HTTP header
func RootHandler(resp http.ResponseWriter, req *http.Request) {
	http.Redirect(resp, req, "/ping", http.StatusSeeOther)
}

// PingHandler handles request to the "/ping" endpoint.
// It will send a PING request to Redis and return the response
// of the NoSQL database.
// The response is obvious: "pong" :)
func PingHandler(s Storage) http.HandlerFunc {
	return func(resp http.ResponseWriter, req *http.Request) {
		res, err := s.Ping()
		if err != nil {
			resp.WriteHeader(http.StatusInternalServerError)
			fmt.Fprintf(resp, err.Error())
			return
		}
		resp.WriteHeader(http.StatusOK)
		fmt.Fprintln(resp, res)
	}
}

// KillHandler handles request to the "/kill" endpoint.
// Will shut down the webserver immediately (via exit code 1).
// Only DELETE requests are accepted.
// Other request methods will throw a HTTP Status Code 405 (Method Not Allowed)
func KillHandler(resp http.ResponseWriter, req *http.Request) {
	if req.Method != "DELETE" {
		resp.WriteHeader(http.StatusMethodNotAllowed)
		return
	}

	// We need to send a HTTP Status Code 200 (OK)
	// to respond that we have accepted the request.
	// Here we send a chunked response to the requester.
	flusher, ok := resp.(http.Flusher)
	if !ok {
		resp.WriteHeader(http.StatusInternalServerError)
		return
	}
	resp.WriteHeader(http.StatusOK)
	flusher.Flush()

	// And we kill the server (like requested)
	os.Exit(1)
}

// VersionHandler handles request to the "/version" endpoint.
// It prints the Name and Version of this app.
func VersionHandler(resp http.ResponseWriter, req *http.Request) {
	resp.WriteHeader(http.StatusOK)
	fmt.Fprintf(resp, "%s v%s\n", Name, Version)
}

// PayloadHandler handles request to the "/payload" endpoint.
// It is a debug route to dump the complete request incl. method, header and body.
func PayloadHandler(resp http.ResponseWriter, req *http.Request) {
	body, err := ioutil.ReadAll(req.Body)
	if err != nil {
		resp.WriteHeader(http.StatusInternalServerError)
		return
	}

	resp.WriteHeader(http.StatusOK)
	log.Printf("Method: %s\n", req.Method)
	fmt.Fprintf(resp, "Method: %s\n", req.Method)

	if len(req.Header) > 0 {
		log.Println("Headers:")
		fmt.Fprint(resp, "Headers:\n")
		for key, values := range req.Header {
			for _, val := range values {
				log.Printf("%s: %s\n", key, val)
				fmt.Fprintf(resp, "%s: %s\n", key, val)
			}
		}
	}

	log.Printf("Payload: %s", string(body))
	fmt.Fprintf(resp, "Payload: %s", string(body))
}

// EnvOrDefault will read env from the environment.
// If the environment variable is not set in the environment
// fallback will be returned.
// This function can be used as a value for flag.String to enable
// env var support for your binary flags.
func EnvOrDefault(env, fallback string) string {
	value := fallback
	if v := os.Getenv(env); v != "" {
		value = v
	}

	return value
}
